curvlinops/kfac.py

```diff
@@ -312,6 +312,90 @@ def _reset_matrix_properties(self):
         self._logdet = None
         self._frobenius_norm = None
 
+    @staticmethod
+    def _left_and_right_multiply(
+        M_joint: Tensor,
+        aaT: FactorType,
+        ggT: FactorType,
+        eigenvalues: Optional[Tensor] = None,
+    ) -> Tensor:
+        """Left and right multiply matrix with Kronecker factors.
+
+        Args:
+            M_joint: Matrix for multiplication.
+            aaT: Input covariance Kronecker factor or its eigenvectors. ``None`` for
+                biases.
+            ggT: Gradient covariance Kronecker factor or its eigenvectors.
+            eigenvalues: Eigenvalues of the (E)KFAC approximation when multiplying with
+                the eigendecomposition of the KFAC approximation. ``None`` for the
+                non-decomposed KFAC approximation. Defaults to ``None``.
+
+        Returns:
+            Matrix-multiplication result ``KFAC @ M_joint``.
+        """
+        if eigenvalues is None:
+            M_joint = einsum(ggT, M_joint, aaT, "i j, j k v, k l -> i l v")
+        else:
+            # Perform preconditioning in KFE, e.g. see equation (21) in
+            # https://arxiv.org/abs/2308.03296.
+            aaT_eigvecs = aaT
+            ggT_eigvecs = ggT
+            # Transform in eigenbasis.
+            M_joint = einsum(
+                ggT_eigvecs, M_joint, aaT_eigvecs, "i j, i k v, k l -> j l v"
+            )
+            # Multiply (broadcasted) by eigenvalues.
+            M_joint.mul_(eigenvalues.unsqueeze(-1))
+            # Transform back to standard basis.
+            M_joint = einsum(
+                ggT_eigvecs, M_joint, aaT_eigvecs, "i j, j k v, l k -> i l v"
+            )
+        return M_joint
```

Comment:
Could rename `M_joint` into `M` since the `_joint` does not add much meaning here.
We should document that `M` is a 3d tensor: the first axis matches the first Kronecker matrix, the second axis the second Kronecker matrix, and the last is the free dimension for carrying out multiple multiplies in parallel.